### PR TITLE
Support earlier "non-QMK" YMD09 keyboard

### PR DESCRIPTION
**Pull request by openfirmware**
## Description

This board predates the YMD09 that officially supports QMK (added in [PR#7831](https://github.com/qmk/qmk_firmware/pull/7831)). Both boards look nearly identical from external appearance; circuits are different after you open them.

I was able to test the VIA keymap with VIA 1.3.1 on MacOS 10.13 and all RGB animation functions seem to work without issue.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by zvecr — CHANGES_REQUESTED** [review]
This should really be within a `ymdk/ymd09/rev1` folder, and the rest of that board migrated to the `rev2` folder where appropriate. `DEFAULT_FOLDER` can then be used to maintain the existing build command.

**Comment by openfirmware** [comment]
I used the `ymd75` keyboard as a template for the organization, hopefully I didn't miss anything!

* created `ymdk/ymd09/rev2` with specifics to the existing board
* moved `ymdk/ymd09v1` to `ymdk/ymd09/rev1`
* set `DEFAULT_FOLDER` to use `ymdk/ymd09/rev2`

File: keyboards/ymdk/ymd09/config.h [hunk 1]
```diff
@@ -20,37 +20,14 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-#define VENDOR_ID 0x594D // "YM"
-#define PRODUCT_ID 0x4409 // "D" + 09
-#define DEVICE_VER 0x0001
+#define VENDOR_ID    0x594D // "YM"
+#define PRODUCT_ID   0x4409 // "D" + 09
 #define MANUFACTURER YMDK
-#define PRODUCT YMD09
-#define DESCRIPTION 9-Key Macropad
+#define PRODUCT      YMD09
+#define DESCRIPTION  9-Key Macropad
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/config.h:27]:
```suggestion
```

File: keyboards/ymdk/ymd09/readme.md [hunk 2]
```diff
@@ -4,12 +4,22 @@
...
+    - [rev2](https://www.aliexpress.com/i/4000510880374.html)
 
 Make example for this keyboard (after setting up your build environment):
 
-    make ymdk/ymd09:default
+    make ymdk/ymd09/rev2:default
+
+For the earlier-release `rev1` board:
+
+    make ymdk/ymd09/rev1:default:bootloadHID
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/readme.md:23]:
```suggestion
    make ymdk/ymd09/rev1:default
    make ymdk/ymd09/rev2:default
```

File: keyboards/ymdk/ymd09/rev1/config.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ #define RGBLIGHT_HUE_STEP 8
+ #define RGBLIGHT_SAT_STEP 8
+ #define RGBLIGHT_VAL_STEP 8
+ #define RGBLIGHT_LIMIT_VAL 255 /* The maximum brightness level */
+ #define RGBLIGHT_SLEEP  /* If defined, the RGB lighting will be switched off when the host goes to sleep */
+ /*== all animations enable ==*/
+ #define RGBLIGHT_ANIMATIONS
+#endif
+
+#define KEYBOARD_ymd09_rev1
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev1/config.h:40]:
```suggestion
```

File: keyboards/ymdk/ymd09/rev1/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+  [0] = LAYOUT(KC_KP_7, KC_KP_8, KC_KP_9,
+               KC_KP_4, KC_KP_5, KC_KP_6,
+               KC_KP_1, KC_KP_2, KC_KP_3),
+
+  [1] = LAYOUT(RGB_RMOD, RGB_VAI, RGB_MOD,
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev1/keymaps/default/keymap.c:25]:
```suggestion
  [0] = LAYOUT_ortho_3x3(KC_KP_7, KC_KP_8, KC_KP_9,
               KC_KP_4, KC_KP_5, KC_KP_6,
               KC_KP_1, KC_KP_2, KC_KP_3),

  [1] = LAYOUT_ortho_3x3(RGB_RMOD, RGB_VAI, RGB_MOD,
```

File: keyboards/ymdk/ymd09/rev1/keymaps/via/keymap.c [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  [1] = LAYOUT(RGB_RMOD, RGB_VAI, RGB_MOD,
+               RGB_HUI,  RESET,   RGB_SAI,
+               RGB_HUD,  RGB_VAD, RGB_SAD),
+
+  [2] = LAYOUT(KC_TRNS, KC_TRNS, KC_TRNS,
+               KC_TRNS, KC_TRNS, KC_TRNS,
+               KC_TRNS, KC_TRNS, KC_TRNS),
+
+  [3] = LAYOUT(KC_TRNS, KC_TRNS, KC_TRNS,
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev1/keymaps/via/keymap.c:33]:
```suggestion
  [0] = LAYOUT_ortho_3x3(KC_KP_7, KC_KP_8,        KC_KP_9,
               KC_KP_4, LT(1, KC_KP_5), KC_KP_6,
               KC_KP_1, KC_KP_2,        KC_KP_3),

  [1] = LAYOUT_ortho_3x3(RGB_RMOD, RGB_VAI, RGB_MOD,
               RGB_HUI,  RESET,   RGB_SAI,
               RGB_HUD,  RGB_VAD, RGB_SAD),

  [2] = LAYOUT_ortho_3x3(KC_TRNS, KC_TRNS, KC_TRNS,
               KC_TRNS, KC_TRNS, KC_TRNS,
               KC_TRNS, KC_TRNS, KC_TRNS),

  [3] = LAYOUT_ortho_3x3(KC_TRNS, KC_TRNS, KC_TRNS,
```

File: keyboards/ymdk/ymd09/rev1/readme.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+A 9-key macropad by [YMDK](https://www.aliexpress.com/store/429151) with per-key RGB lighting. [This is an earlier board](https://www.aliexpress.com/item/33008853928.html) than the [YMD09 that officially supports QMK](https://www.aliexpress.com/item/4000510880374.html).
+
+* Keyboard Maintainer: [James Badger](https://github.com/openfirmware)
+* Hardware Supported:
+    - rev1 ("non-QMK version")
+* Hardware Availability: [AliExpress](https://www.aliexpress.com/item/33008853928.html)
+
+Make example for this keyboard (after setting up your build environment):
+
+    make ymdk/ymd09/rev1:default:bootloadHID
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev1/readme.md:14]:
```suggestion
    make ymdk/ymd09/rev1:default
```

File: keyboards/ymdk/ymd09/rev1/rev1.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev1/rev1.h:20]:
```suggestion
#define LAYOUT_ortho_3x3( \
```

File: keyboards/ymdk/ymd09/rev2/config.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ #define RGBLIGHT_HUE_STEP 8
+ #define RGBLIGHT_SAT_STEP 8
+ #define RGBLIGHT_VAL_STEP 8
+ #define RGBLIGHT_LIMIT_VAL 255 /* The maximum brightness level */
+ #define RGBLIGHT_SLEEP  /* If defined, the RGB lighting will be switched off when the host goes to sleep */
+ /*== all animations enable ==*/
+ #define RGBLIGHT_ANIMATIONS
+#endif
+
+#define KEYBOARD_ymd09_rev2
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev2/config.h:23]:
```suggestion
```

File: keyboards/ymdk/ymd09/rev2/rules.mk [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev2/rules.mk:11]:
```suggestion
```

File: keyboards/ymdk/ymd09/rev2/rules.mk [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
+LTO_ENABLE = yes            # reduce firmware size
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev2/rules.mk:33]:
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
LTO_ENABLE = yes            # reduce firmware size
```

File: keyboards/ymdk/ymd09/rev2/rev2.h [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev2/rev2.h:4]:
```suggestion
#define LAYOUT_ortho_3x3( \
```

File: keyboards/ymdk/ymd09/rev1/rules.mk [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = no            # USB Nkey Rollover
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+WS2812_DRIVER = i2c
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/rev1/rules.mk:17]:
```suggestion
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
WS2812_DRIVER = i2c
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by zvecr — CHANGES_REQUESTED** [review]
Unless I have missed something, you can avoid having keymap folders within each revision. A single `keyboards/ymdk/ymd09/keymaps` folder will avoid it being considered a "breaking change".

**Comment by openfirmware** [comment]
My apologies for the late response, I plan to spend some time this week updating the PR. Thanks for giving the suggestions and changes, I am still new to QMK!

**Comment by openfirmware** [comment]
I had some build issues after merging the changes; I found more occurrences of `LAYOUT` in the base keymaps directory that I replaced with `LAYOUT_ortho_3x3`, as well as in the `info.json` file. It still wouldn't build, but if I re-add the `#define KEYBOARD_ymd09_rev1` and `#define KEYBOARD_ymd09_rev2` to the `config.h` files, it builds.

Is there a better way of including revision-specific header files than what is going on in the base `ymd09.h` file?

```c
#if defined(KEYBOARD_ymd09_rev1)
    #include "rev1.h"
#elif defined(KEYBOARD_ymd09_rev2)
    #include "rev2.h"
#endif
```

As a side note, I noticed the board under some circumstances will lag (run extremely slow) while animating the LEDs, possibly indicating some sort of misconfiguration. This was before I merged in the changes from @fauxpark, so I will keep an eye out to see if it is still happening.

File: keyboards/ymdk/ymd09/ymd09.h [hunk 13]
```diff
@@ -2,12 +2,8 @@
...
-	K10, K11, K12, \
-	K20, K21, K22  \
-) { \
-	{ K00,   K01,   K02 }, \
-	{ K10,   K11,   K12 }, \
-	{ K20,   K21,   K22 }  \
-}
+#if defined(KEYBOARD_ymd09_rev1)
+    #include "rev1.h"
+#elif defined(KEYBOARD_ymd09_rev2)
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/ymd09.h:7]:
```suggestion
#if defined(KEYBOARD_ymdk_ymd09_rev1)
    #include "rev1.h"
#elif defined(KEYBOARD_ymdk_ymd09_rev2)
```

File: keyboards/ymdk/ymd09/info.json [hunk 14]
```diff
@@ -5,7 +5,7 @@
     "width": 3,
     "height": 3,
     "layouts": {
-        "LAYOUT": {
+        "LAYOUT_ortho_3x3": {
             "key_count": 9,
```
Comment (fauxpark) [inline keyboards/ymdk/ymd09/info.json:9]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/ymdk/ymd09/rev2/rev2.h [hunk 15]
```diff
@@ -0,0 +1,12 @@
+#pragma once
```
Comment (drashna) [inline keyboards/ymdk/ymd09/rev2/rev2.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/ymdk/ymd09/rev2/rev2.h [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+
+#define LAYOUT_ortho_3x3( \
+    K00, K01, K02, \
+    K10, K11, K12, \
+    K20, K21, K22  \
+) { \
+    { K00,   K01,   K02 }, \
+    { K10,   K11,   K12 }, \
+    { K20,   K21,   K22 }  \
+}
```
Comment (zvecr) [inline keyboards/ymdk/ymd09/rev2/rev2.h:12]: For backwards compatibility, 

```suggestion
}

#define LAYOUT LAYOUT_ortho_3x3
```

**Review by zvecr — CHANGES_REQUESTED** [review]
> Could you add a GPL2+ compatible license header here?

As with the above request, this should apply to the added `*.h` and `*.c` files.

File: keyboards/ymdk/ymd09/rev2/rev2.c [hunk 17]
```diff
@@ -0,0 +1 @@
+#include "ymd09.h"
```
Comment (drashna) [inline keyboards/ymdk/ymd09/rev2/rev2.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/ymdk/ymd09/rev2/config.h [hunk 18]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (drashna) [inline keyboards/ymdk/ymd09/rev2/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — CHANGES_REQUESTED** [review]
You have some merge conflicts here.  

You can update these through the UI (below), or by updating your branch.

**Comment by peepeetee** [comment]
um I guess I'm here to bump this? stumbled upon this while collecting board move information

**Pull request closed without merging**